src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -163,23 +163,25 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
             return;
         }
 
-        Map<Id, npe03__Recurring_Donation__c> oldRdsById = triggerAction == TDTM_Runnable.Action.BeforeUpdate
-            ? new Map<Id, npe03__Recurring_Donation__c>(oldRds)
-            : new Map<Id, npe03__Recurring_Donation__c>();
-
-        Map<Id, List<RecurringDonationSchedule__c>> schedulesByRdId = triggerAction == TDTM_Runnable.Action.BeforeUpdate
-            ? scheduleService.getAllSchedules(rds, oldRds)
-            : new Map<Id, List<RecurringDonationSchedule__c>>();
-
-        for (npe03__Recurring_Donation__c rd : rds) {
-            RD2_RecurringDonation rdRecord = new RD2_RecurringDonation(rd)
-                .revisePlannedInstallments();
+        Map<Id, npe03__Recurring_Donation__c> oldRdsById = new Map<Id, npe03__Recurring_Donation__c>();
+        Map<Id, List<Opportunity>> oppsByRDId = new Map<Id, List<Opportunity>>();
+        Map<Id, List<RecurringDonationSchedule__c>> schedulesByRdId = new Map<Id, List<RecurringDonationSchedule__c>>();
+        
+        if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+            oldRdsById = new Map<Id, npe03__Recurring_Donation__c>(oldRds);
+            oppsByRDId = new RD2_OpportunityEvaluationService().getOpportunitiesByRDId(UTIL_SObject.extractIds(rds));
+            schedulesByRdId = scheduleService.getAllSchedules(rds, oldRds);
+        }
 
+        for (npe03__Recurring_Donation__c rd : rds) { 
             if (triggerAction == TDTM_Runnable.Action.BeforeInsert) {
-                rdRecord.reviseNextDonationDateBeforeInsert(scheduleService);
+                RD2_RecurringDonation rdRecord = new RD2_RecurringDonation(rd)
+                    .revisePlannedInstallments()
+                    .reviseNextDonationDateBeforeInsert(scheduleService);
 
-            } else if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
-                rdRecord
+            } else if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {                
+                RD2_RecurringDonation rdRecord = new RD2_RecurringDonation(rd, oppsByRDId.get(rd.Id))
+                    .revisePlannedInstallments()
                     .reviseStatus(currentDate)
                     .reviseNextDonationDate(scheduleService, schedulesByRdId.get(rd.Id));
```

Comment:
Even when an RD has a closed Opportunity for the first future installment, the Next Donation Date would be set to the first installment date in the RD2 trigger on update, since Opportunities were not retrieved for the NDD revision so the first installment was considered as open. However, the Opp eval service would overwrite the value relatively soon since the service is queued from the RD2 trigger. So now the Opportunities are fetched from `RD2_OpportunityEvaluationService().getOpportunitiesByRDId()`.

The validation service also fetches Opportunities, so we added one more SOQL in the transaction. However, I think it is better to keep responsibilities of fetching Opps separately.